### PR TITLE
✨  [amp-animation] Adds a new attr that allows to override pan scaling factor.

### LINKED ISSUE #37932: [amp-story] Provide an option in animation preset to disable image scaling for pan effects.
### Description

The amp-story animation preset scales the pan effect image. For example, `pan-right` effect:

https://github.com/ampproject/amphtml/blob/8c34f2d8d974eba8e926f06f400be482797e9d0a/extensions/amp-story/1.0/animation-presets.js#L304-L325

The `calculateTargetScalingFactor` method ensures the image is at least 25% larger than the page; however, the logic is not useful when the image is not full-bleed in the page. For example, a small image in the story page with pan- animation might be scaled to 2x while the expectation is not to scale.

Hope to provide an option such as `animation-scale-disabled` to disable the scaling.

https://github.com/ampproject/amphtml/blob/8c34f2d8d974eba8e926f06f400be482797e9d0a/extensions/amp-story/1.0/animation-presets-utils.js#L76-L82

/cc @newmuis 

### Alternatives Considered

Use other animation, or create a custom panning animation. Yet to resolve it, providing an option to disable the scaling would help the most.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by jingfei**
Currently amp-story animation preset scales the pan effect image by checking if the target is smaller than the story page; however, for the target that is not designed as full-bleed, it causes an unexpected scaling. For example, a small image in the story page with pan- animation might be scaled to 2x while the expectation is not to scale.

The PR provides a new attribution, `pan-static-scale`, that allows the pan animation to override the scaling factor with a static value, and it would fixes #37932.

/cc @newmuis

### CODE REVIEW

File: extensions/amp-story/1.0/animation-presets.js [hunk 1]
```diff
@@ -23,6 +23,8 @@ const SCALE_START_ATTRIBUTE_NAME = 'scale-start';
 /** @const {string} */
 const SCALE_END_ATTRIBUTE_NAME = 'scale-end';
 /** @const {string} */
+const PAN_STATIC_SCALE_ATTRIBUTION_NAME = 'pan-static-scale';
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/animation-presets.js:26]: Here and below:

```suggestion
const PAN_STATIC_SCALE_ATTRIBUTE_NAME = 'pan-static-scale';
```

**Review by gmajoulet — COMMENTED** [review]
Thank you for opening a PR! Could you post an example where the image is scaled 2x? I'd like to understand how bad the issue can get.
The one thing I'm worried about is that most publishers have historically not been able to test their layouts/animations on enough devices to get them to be always displayed correctly. I'm afraid for us to open this setting and publishers getting it wrong for some devices.

**Comment by jingfei** [comment]
Sure no problem. As long as the image is not full-bleed it will be scaled to fit the story page if pan effect is added.

Here's a quick example, the 3 pages in the story are the same except for pan animation
- page 1: no animation
- page 2: pan-right effect --> it scales the image to be same as the page width/height, which is not expected in our use case
- page 3: zoom-in effect --> as the scaling factor is provided by user, it works as expected in our use case.

With the new added attr, it's able to be animated with pan effect and custom scaling factor.

Let me know if you have questions, happy to discuss more in details. Thank you.

```html
<!doctype html>
<html ⚡ lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width">
    <meta name="description" content="This is the AMP Boilerplate.">
    <link rel="preload" as="script" href="https://cdn.ampproject.org/v0.js">
    <link rel="preload" as="script" href="https://cdn.ampproject.org/v0/amp-story-1.0.js">
    <script async src="https://cdn.ampproject.org/v0.js"></script>
    <script async custom-element="amp-story" src="https://cdn.ampproject.org/v0/amp-story-1.0.js"></script>
    <!-- Import other AMP Extensions here -->
    <style amp-custom>
    /* Add your styles here */
    </style>
    <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
    <link rel="canonical" href=".">
    <title>My AMP Page</title>
  </head>
  <body>
    <amp-story standalone
               publisher="The Publisher"
               title="My Story"
               publisher-logo-src="logo.png"
               poster-portrait-src="poster-portrait.png"
               poster-square-src="poster-square.png"
               poster-landscape-src="poster-landscape.png">
      <amp-story-page id="page1">
        <amp-story-grid-layer template="vertical">
          <h1>Hello World</h1>
          <div style="width:420px;height:315px;overflow:hidden">
            <amp-img layout="responsive" height="960" width="1280" src="https://upload.wikimedia.org/wikipedia/commons/thumb/b/b6/Image_created_with_a_mobile_phone.png/1280px-Image_created_with_a_mobile_phone.png" />
          </div>
        </amp-story-grid-layer>
      </amp-story-page>
      <amp-story-page id="page2">
        <amp-story-grid-layer template="vertical">
          <h1>Hello Animation</h1>
          <div style="width:420px;height:315px;overflow:hidden">
            <amp-img animate-in="pan-right" layout="responsive" height="960" width="1280" src="https://upload.wikimedia.org/wikipedia/commons/thumb/b/b6/Image_created_with_a_mobile_phone.png/1280px-Image_created_with_a_mobile_phone.png" />
          </div>
        </amp-story-grid-layer>
      </amp-story-page>
      <amp-story-page id="page3">
        <amp-story-grid-layer template="vertical">
          <div style="width:420px;height:315px;overflow:hidden">
          <amp-img animate-in="zoom-in" layout="responsive" height="960" width="1280" src="https://upload.wikimedia.org/wikipedia/commons/thumb/b/b6/Image_created_with_a_mobile_phone.png/1280px-Image_created_with_a_mobile_phone.png" />
          </div>
        </amp-story-grid-layer>
      </amp-story-page>
      <amp-story-bookend src="bookend.json" layout=nodisplay></amp-story-bookend>
    </amp-story>
  </body>
</html>
```

File: extensions/amp-story/1.0/animation-presets.js (hunk 1 above)
Comment (jingfei) [reply extensions/amp-story/1.0/animation-presets.js]: Updated.

**Comment by gmajoulet** [comment]
cc @mszylkowski can you make sure this is the right fix to this problem and review? Thank you!

**Comment by gmajoulet** [comment]
Also, please don't forget to udpate the validation rules so you can add the attribute without making your stories invalid:

https://github.com/ampproject/amphtml/blob/bccc2a374f98136037f33b453664934bf412bf38/extensions/amp-story/validator-amp-story.protoascii#L222-L248

**Comment by jingfei** [comment]
No problem! Validation rule added.
Thanks for the reminder.

**Review by mszylkowski — COMMENTED** [review]
I think Gabriel commented before, but it's super hard for creators to know how this attribute will behave for different page sizes (it's likely that 90% of that comes from how hard it is to understand the current logic).

There's an issue with this logic, which is that the images by default have pixel sizes (the value on `width` and `height`) and the scaling factor helps to correct that, although it's really bad at it.

The correct way to fix this would be to check the size of the element instead of the page, so that smaller containers will not require such large zoom levels. The fast way to fix this is to multiply the scaling factor calculated by the attribute that you want to pass here, so that there's still smarter scaling, but you can specify the "ratio" that you want to use (eg: 0.5 for an element that occupies 50% of the page)

File: extensions/amp-story/validator-amp-story.protoascii [hunk 2]
```diff
@@ -267,6 +267,10 @@ tags: {
     name: "translate-y"
     value_regex_casei: "[0-9]+px"
   }
+  attrs: {
+    name: "pan-static-scale"
+    value_regex_casei: "[0-9]+([.][0-9]+)?"
```
Comment (MichaelRybak) [inline extensions/amp-story/validator-amp-story.protoascii:272]: Hi, I'll review the validation changes on behalf of wg-caching group.

Should the regex also allow a decimal point with no fractional part or no integer part, e.g. "123." and ".2"? E.g. css scale transform allows both (but not both at the same time).

File: extensions/amp-story/1.0/animation-presets.js [hunk 3]
```diff
@@ -349,7 +357,9 @@ export const presets = {
     easing: 'linear',
     keyframes(dimensions, options) {
       const translateY = options[TRANSLATE_Y_ATTRIBUTE_NAME];
-      const scalingFactor = calculateTargetScalingFactor(dimensions);
+      const scalingFactor = options[PAN_STATIC_SCALE_ATTRIBUTE_NAME]
+        ? options[PAN_STATIC_SCALE_ATTRIBUTE_NAME]
```
[the remark this answers is not in the document]
Comment (jingfei) [reply extensions/amp-story/1.0/animation-presets.js:361]: Sure, updated. TIL!

File: extensions/amp-story/validator-amp-story.protoascii (hunk 2 above)
Comment (jingfei) [reply extensions/amp-story/validator-amp-story.protoascii:272]: No problem, done.

**Comment by jingfei** [comment]
As discussed with @mszylkowski offline, will keep the CL as-is to apply to our use case. At the same time, no further CSS change (e.g. to remove full-bleed CSS) will make to reduce the additional complication.

Also, in the latest changes, I updated the validation rules with tests. 

PTAL. Thank you.

File: extensions/amp-story/1.0/test/test-full-bleed-animations.js [hunk 4]
```diff
@@ -248,7 +247,45 @@ describes.realWin(
...
-      expect(calculatedKeyframes.keyframes).to.deep.equal(expectedKeyframes);
+      expect(calculatedKeyframes).to.deep.equal(expectedKeyframes);
+    });
+
+    ['pan-up', 'pan-down', 'pan-right', 'pan-left'].forEach((panAnimation) => {
+      it(`Should scale the target for ${panAnimation}.`, () => {
+        const dimensions = setDimensions(380, 580, 360, 580);
+        expect(targetFitsWithinPage(dimensions)).to.be.true;
+
+        const factorThatWillMakeTargetFitPage = 380 / 360;
```
[the remark this answers is not in the document]
Comment (jingfei) [reply extensions/amp-story/1.0/test/test-full-bleed-animations.js:258]: +1, updated.

**Review by MichaelRybak — APPROVED** [review]
Approving validator changes.

**Review by mszylkowski — APPROVED** [review]

**Pull request merged**